src/main/java/com/google/step/servlets/HomePageServlet.java

```diff
@@ -0,0 +1,219 @@
+import com.google.step.datamanager.TagManager;
+import com.google.step.datamanager.TagManagerDatastore;
+import com.google.step.datamanager.UserManager;
+import com.google.step.datamanager.UserManagerDatastore;
+import com.google.step.datamanager.VoteManager;
+import com.google.step.datamanager.VoteManagerDatastore;
+import com.google.step.model.Deal;
+import com.google.step.model.Restaurant;
+import com.google.step.model.Tag;
+import com.google.step.model.User;
+import java.io.IOException;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that handles retrieving deals for home page */
+@WebServlet("/api/home")
+public class HomePageServlet extends HttpServlet {
+
+  private final DealManager dealManager;
+  private final UserManager userManager;
+  private final VoteManager voteManager;
+  private final RestaurantManager restaurantManager;
+  private final DealTagManager dealTagManager;
+  private final TagManager tagManager;
+  private final FollowManager followManager;
+
+  private final Long OLDEST_DEAL_TIMESTAMP = 1594652120L; // arbitrary datetime of first deal posted
+  private final String LOCATION = "Asia/Singapore";
+
+  public HomePageServlet(
+      DealManager dealManager,
+      UserManager userManager,
+      RestaurantManager restaurantManager,
+      VoteManager voteManager,
+      DealTagManager dealTagManager,
+      TagManager tagManager,
+      FollowManager followManager) {
+    this.dealManager = dealManager;
+    this.userManager = userManager;
+    this.voteManager = voteManager;
+    this.restaurantManager = restaurantManager;
+    this.dealTagManager = dealTagManager;
+    this.tagManager = tagManager;
+    this.followManager = followManager;
+  }
+
+  public HomePageServlet() {
+    dealManager = new DealManagerDatastore();
+    userManager = new UserManagerDatastore();
+    voteManager = new VoteManagerDatastore();
+    restaurantManager = new RestaurantManagerDatastore();
+    tagManager = new TagManagerDatastore();
+    dealTagManager = new DealTagManagerDatastore();
+    followManager = new FollowManagerDatastore();
+  }
+
+  /** Class to store deal along with relevant attribute (hot score/votes) to be sorted */
+  class DealPair<T extends Comparable<T>> implements Comparable<DealPair<T>> {
+    public final T key;
+    public final Deal deal;
+
+    public DealPair(T key, Deal deal) {
+      this.key = key;
+      this.deal = deal;
+    }
+
+    @Override
+    public int compareTo(DealPair<T> other) {
+      return key.compareTo(other.key);
+    }
+  }
```

Comment:
This class doesn't need to use generics and can be named better.

```java
class ScoredDeal implements Comparable<ScoredDeal> {
  public final Deal deal;
  public final double score;

  public ScoredDeal(Deal deal, double score) { /* this can accept int or double scores */ }

  @Override
  public int compareTo(ScoredDeal<T> other) {}
}
```